latte/dataset/dsprites.py

```diff
@@ -125,6 +164,132 @@ def load_dsprites(filepath: Union[str, pathlib.Path] = DEFAULT_TARGET, download_
     )
 
 
+def prepare_data(
+    data: DSpritesDataset, factors: List[DSpritesFactor], ignored_factors: List[DSpritesFactor]
+) -> Tuple[np.ndarray, np.ndarray, Dict[DSpritesFactor, int]]:
+    """Converts the raw dSprites data into Numpy arrays which will be
+    processed by PCA.
+    Args:
+        data: DSpritesDataset object
+        factors: Which factors to keep the information about in the returned data
+        ignored_factors: Which factors should be held fixed (all the variation in them will be removed)
+
+    Returns:
+        Flattened filtered images and factors
+
+    """
+    # Flatten images
+    X = data.imgs.reshape((len(data.imgs), -1))
+
+    # Select the images of interest by ignoring the variation of factors
+    # we are not interested in
+    sample_mask = np.ones(data.latents_values.shape[0], dtype=bool)
+    for factor in ignored_factors:
+        sample_mask *= data.latents_values[:, factor] == fixed_values[DSpritesFactor(factor)]
+
+    X = X[sample_mask, :]
+    y = data.latents_values[sample_mask, :]
+
+    # Filter out constant pixels
+    X = X[:, X.max(axis=0) == 1]
+
+    # Keep only the factors that we want
+    y = y[:, factors]
+
+    # The dictionary which maps each of the factors of interest
+    # into the dimension of the returned factors y
+    factor2idx = {factor: ii for ii, factor in enumerate(factors)}
+
+    return X, y, factor2idx
+
+
+def load_dsprites_preprocessed(
+    filepath: Union[str, pathlib.Path] = DEFAULT_TARGET,
+    download_ok: bool = True,
+    method: str = "PCA",
+    n_components: int = 20,
+    factors: Optional[List[DSpritesFactor]] = None,
+    ignored_factors: Optional[List[DSpritesFactor]] = None,
+) -> PreprocessedDSpritesDataset:
+    """Load the dSprites dataset and preprocess it using PCA or FactorAnalysis.
+    Args:
+        filepath: The location of the dsprites data.
+        download_ok: If the data can be downloaded.
+        method: The way to preprocess the data, {PCA, FA]
+        n_components: Number of component to produce in the representation.
+        factors: Which factors to return.
+        ignored_factors: Which factors should be kept constant in the dataset.
+            The data is filtered and reduced such that these factors have a constant value.
+
+    Returns:
+        A preprocessedDSpritesDataset holding the filtered transformed data
+        (coordinates in PCA or FA component space), the filtered latent factors,
+        the PCA variance explained proportions in case method="PCA", and a dictionary
+        mapping each of the returned factors to the dimension in `y` to be able to select
+        appropriate factors from the returned ones.
+    """
+
+    assert method in [
+        "PCA",
+        "FA",
+        "FactorAnalysis",
+    ], f"Method {method} is not supported, only PCA and FactorAnalysis supported."
+
+    if factors is None:
+        # By default, return all factors
+        factors = [
```

Comment:
I think we can do this in a simpler way – what do you think about `list(factor_names.keys())`?